File: investment_screener/backend/src/utils/tickerAliases.ts
/**
 * tickerAliases.ts - Canonical ticker symbol normalization.
 * 
 * Purpose:
 *   Normalizes broker-format symbols (e.g. PSU.U.TO) to canonical Yahoo Finance/TSX formats
 *   (e.g. PSU-U.TO) to prevent duplicate holdings rows. Keep in sync with py_services/ticker_aliases.py.
 * 
 * Key Input Dependencies:
 *   None
 * 
 * Key Output Dependencies:
 *   None
 * 
 * Functions Index:
 *   - normalizeTicker(ticker: string) - Return the canonical ticker resolving broker aliases
 */

export const TICKER_ALIASES: Record<string, string> = {
    'PSU.U': 'PSU-U.TO',
    'PSU.U.TO': 'PSU-U.TO',
};

/**
 * Return the canonical ticker, resolving any broker aliases.
 * 
 * @param {string} ticker - Broker symbol ticker string
 * @returns {string} Canonical resolved ticker string
 */
export function normalizeTicker(ticker: string): string {
    /**
     * Looks up ticker in aliases hash map and returns the mapped canonical name or the input string on fallback.
     */
    return TICKER_ALIASES[ticker] ?? ticker;
}

File: investment_screener/backend/py_services/ticker_aliases.py
"""
ticker_aliases.py — Canonical ticker normalization and special-symbol definitions.

Single source of truth for:
  - Ticker aliases (e.g. PSU.U → PSU-U.TO from the Questrade broker format)
  - Reserved cash symbols (USD_CASH — the internal sentinel for uninvested USD)

Import this module instead of scattering inline `if ticker == "USD_CASH"` checks.

Key Input Dependencies:
    - investment_screener/backend/data/portfolio.json (Maps ticker synonyms)
"""

# Questrade returns "PSU.U" without the exchange suffix; normalize to the TSX symbol.
# Add entries here when a new broker alias is discovered.
TICKER_ALIASES: dict[str, str] = {
    "PSU.U": "PSU-U.TO",
    "PSU.U.TO": "PSU-U.TO",
}

# Tickers that represent cash/liquidity reserves, not tradeable equities.
# These are excluded from valuation, drift, and opportunity scans.
CASH_SYMBOLS: frozenset[str] = frozenset({"USD_CASH"})

# Human-readable display names for reserved symbols.
CASH_DISPLAY_NAMES: dict[str, str] = {
    "USD_CASH": "US Dollar Cash",
}


def normalize_ticker(ticker: str) -> str:
    """Return the canonical ticker, resolving any broker aliases."""
    return TICKER_ALIASES.get(ticker, ticker)


def is_cash(ticker: str) -> bool:
    """Return True if the ticker represents a cash/liquidity position."""
    return ticker in CASH_SYMBOLS or ticker.endswith("_CASH")
